compiler/src/vm/tracer/stack_trace.rs

```diff
@@ -0,0 +1,184 @@
+use super::{super::heap::Pointer, Event, FiberId, FullTracer, InFiberEvent};
+use crate::{
+    compiler::{
+        ast_to_hir::AstToHir,
+        cst::{Cst, CstDb, CstKind},
+        hir::Id,
+    },
+    database::Database,
+    language_server::utils::LspPositionConversion,
+    module::Module,
+};
+use itertools::Itertools;
+use std::collections::HashMap;
+
+// Stack traces are a reduced view of the tracing state that represent the stack
+// trace at a given moment in time.
+
+#[derive(Clone)]
+pub enum StackEntry {
+    Call {
+        id: Id,
+        closure: Pointer,
+        args: Vec<Pointer>,
+    },
+    Needs {
+        id: Id,
+        condition: Pointer,
+        reason: Pointer,
+    },
+    Module {
+        module: Module,
+    },
+}
+
+impl FullTracer {
+    pub fn stack_traces(&self) -> HashMap<FiberId, Vec<StackEntry>> {
+        let mut stacks: HashMap<FiberId, Vec<StackEntry>> = HashMap::new();
+        for timed_event in &self.events {
+            if let Event::InFiber { fiber, event } = &timed_event.event {
+                let stack = stacks.entry(*fiber).or_default();
+                match event {
+                    InFiberEvent::ModuleStarted { module } => {
+                        stack.push(StackEntry::Module {
+                            module: module.clone(),
+                        });
+                    }
+                    InFiberEvent::ModuleEnded { .. } => {
+                        stack.pop().unwrap();
```

Comment:
The `…Ended` cases could assert that the popped entry is of the same type, but it's not necessary